lib/block_view_validator.go

```diff
@@ -0,0 +1,2133 @@
+package lib
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"github.com/deso-protocol/core/bls"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/golang/glog"
+	"github.com/holiman/uint256"
+	"github.com/pkg/errors"
+	"io"
+	"math"
+	"net/url"
+	"sort"
+)
+
+//
+// TYPES: ValidatorEntry
+//
+
+type ValidatorEntry struct {
+	ValidatorID   *BlockHash
+	ValidatorPKID *PKID
+	// Note: if someone is updating their ValidatorEntry, they need to include
+	// all domains. The Domains field is not appended to. It is overwritten.
+	Domains                    [][]byte
+	DisableDelegatedStake      bool
+	VotingPublicKey            *bls.PublicKey
+	VotingPublicKeySignature   *bls.Signature
+	VotingSignatureBlockHeight uint64
+	TotalStakeAmountNanos      *uint256.Int
+	RegisteredAtBlockHeight    uint64
+	LastActiveAtEpochNumber    uint64
+	JailedAtEpochNumber        uint64
+	ExtraData                  map[string][]byte
+	isDeleted                  bool
+}
+
+func (validatorEntry *ValidatorEntry) Status() ValidatorStatus {
+	// ValidatorEntry.Status() is a virtual/derived field that is not stored in
+	// the database, but instead constructed from other ValidatorEntry fields.
+	// No sense in storing duplicative data twice. This saves memory and ensures
+	// that e.g. the ValidatorEntry.JailedAtEpochNumber field and the
+	// ValidatorEntry.Status() return value will never get out of sync.
+	//
+	// Make sure that any fields referenced here are included in the ValidatorMapKey
+	// since the ValidatorEntry.Status() value is used as a field in a Badger index.
+	if validatorEntry.JailedAtEpochNumber > uint64(0) {
+		return ValidatorStatusJailed
+	}
+	return ValidatorStatusActive
+}
+
+type ValidatorStatus uint8
+
+const (
+	ValidatorStatusInvalid ValidatorStatus = 0
+	ValidatorStatusActive  ValidatorStatus = 1
+	ValidatorStatusJailed  ValidatorStatus = 2
+)
+
+type ValidatorMapKey struct {
+	// The MapKey has to contain all fields that are used in Badger keys.
+	// Otherwise, an update to the UtxoView will not be able to update or
+	// delete all relevant Badger rows.
```

Comment:
Ok so I don't think you need to change this necessarily, but I want you to noodle on sometihng.

For PKIDToProfileEntry, we map a PKID->ProfileEntry in the view, even though there are a lot of ancillary indexes.

The reason this works is we do the following trick:
* In the view, if we update a profile, then we'll have PKID->NewProfile.
* When we flush, we call this interesting function called DBDeleteProfileEntryMappingsWithTxn(pkid *PKID)
* What that function does is it first LOOKS UP the ProfileEntry *in the database* that matches the PKID, and then it *uses the DB ProfileEntry* to kill all the mappings in the DB. This works because the ProfileEntry in the DB has all the old values, which tells you how to delete it.
* Then, only after all the mappings are nuked, we call DBPutProfileEntryMappings with the new PKID.

I haven't gone deep enough to know if this would work here, but the way you would do it is you would have a DBDeleteValidatorEntryMappings(validatorPKID *PKID), which you call in a view flush before writing the new ValidatorEntry in the view.

Reply:
Btw the benefit of this is that the view code gets a lot simpler. You're effectively "pushing down" the complexity into the DB layer, which is really where you ideally want it (as low as possible).

Reply:
I fixed this for you and fixed the tests for fun. It wasn't hard, and I think it makes the view a lot cleaner.

Btw you did a great job with the tests. They caught some good stuff when I was making the change :)